app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabase.kt

```diff
@@ -582,26 +636,163 @@ class FirebaseDatabase(reference: DatabaseReference = Firebase.database.referenc
-                tournamentsRoot().child(tournamentId).get().addOnSuccessListener {
-                    val tournament = FirebaseDatabaseUtils.mapToTournament(it)
+                // Get tournament posts
+                tournamentsRoot().child(tournamentId)
+                    .child(FirebaseKeys.TOURNAMENT_PARTICIPANTS_IDS).get().addOnSuccessListener {
+                        future.complete(FirebaseDatabaseUtils.getKeys(it))
+                    }.addOnFailureListener {
+                        future.completeExceptionally(it)
+                    }
+            }
+        }
+        return future
+    }
+
+    override fun getTournamentInfo(tournamentId: String): LiveData<Tournament> {
+        val tournamentInfoValue = MutableLiveData<Tournament>()
+        val tournamentInfoValueEventListener: ValueEventListener
+
+        // listener for data change
+        tournamentInfoValueEventListener = object : ValueEventListener {
+            override fun onCancelled(error: DatabaseError) {
+                throw Exception("The tournament with tournamentId $tournamentId can't be fetched from the database.")
+            }
+
+            override fun onDataChange(snapshot: DataSnapshot) {
+                // Check that the tournament exists
+                if (snapshot.exists()) {
+                    // Get tournament info
+                    val tournament = FirebaseDatabaseUtils.mapToTournamentInfo(snapshot)
                     if (tournament == null) {
-                        future.completeExceptionally(Exception("Corrupted data for the tournament $tournamentId."))
+                        throw Exception("Corrupted data for the tournament $tournamentId.")
                     } else {
-                        future.complete(tournament)
+                        tournamentInfoValue.postValue(tournament!!)
                     }
-                }.addOnFailureListener {
-                    future.completeExceptionally(it)
+                } else {
+                    throw Exception("The tournament with tournamentId $tournamentId doesn't exist.")
                 }
             }
         }
+        // add the listener on the database on the root of the tournament info with tournament id
+        tournamentsRoot().child(tournamentId).child(FirebaseKeys.TOURNAMENT_INFO).ref.addValueEventListener(tournamentInfoValueEventListener)
+        return tournamentInfoValue
+    }
+
+    override fun getPostUID(): String? {
+        return database.child(FirebaseKeys.TOURNAMENTS_ROOT).child(FirebaseKeys.TOURNAMENT_POSTS)
+            .push().key
+    }
+
+    override fun addPostToTournament(
+        tournamentId: String,
+        post: TournamentPost,
+    ): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+        // Check that the tournament exists
+        isTournamentInDatabase(tournamentId).thenApply { tournamentExists ->
+            if (!tournamentExists) {
```

Comment:
Suggestion : Maybe you can create an helper function that directly return a future that complete exceptionally with error to test if the tournament exist.

Reply:
I tried to but I cannot manage to do it in a good way because of the "lambda"... maybe we can let it like that, it's basically just an if-else pre-condition, but a bit more complicated, and pretty hard to modularize (for me at least).

Reply:
No problem, I understand.